In [95]:
from IPython.display import Markdown, display

In [96]:
B0 = 3

Wn = 1
Wp = B0 * Wn # 3

# Static inverter

When they talk about pMOS-to-nMOS ratio $B_0$, it is found by the formula:

$$
B_0 = \frac{W_p}{W_n}
$$

Finding the load for the inverter is

$$
C_{inv} = W_p + W_n
$$

It represents the gate capacitance of an one nMOS gate and one pMOS gate

In [97]:
C_inv = Wp + Wn

latex_cinv = fr"C_{{inv}} = W_p + W_n = {Wp} + {Wn} = {C_inv}"
display(Markdown("$" + latex_cinv + "$"))
print(latex_cinv)

$C_{inv} = W_p + W_n = 3 + 1 = 4$

C_{inv} = W_p + W_n = 3 + 1 = 4


Inverter with ratio 3

<img src="img/invr3.svg" width="10%">

The unit inverter logical effort is by definition:

$$
g_{inv} = 1
$$

The parasitic delay is by definition:

$$
p_{inv} = 1
$$


## n-input NAND gate

A NAND gate has:

- nMOS in series 
- pMOS in parallel

$$
W_{n,NAND} = n \cdot W_n
$$

$$
W_{p,NAND} = W_p
$$

Each input drives one pMOS and one nMOS gate.

$$
C_{in,NAND} = W_{n,NAND} + W_{p,NAND} = n \cdot W_n + W_p
$$

In [98]:
latex_cin_nand = fr"C_{{in,NAND}} = n \cdot W_n + W_p = n \cdot {Wn} + {Wp} = n + {Wp}"
display(Markdown("$" + latex_cin_nand + "$"))
print(latex_cin_nand)

$C_{in,NAND} = n \cdot W_n + W_p = n \cdot 1 + 3 = n + 3$

C_{in,NAND} = n \cdot W_n + W_p = n \cdot 1 + 3 = n + 3


$NAND_n$ ratio 3

<img src="img/nandnr3.svg" width="20%">

Example of a NAND3 with ratio 3.

$W_{n,NAND_3} = n \cdot W_n = 3 \cdot 1 = 3$

$W_{p,NAND_3} = W_p = 3$.

<img src="img/nand3r3.svg" width="20%">

The logical effort is

$$
g_{NAND_{n}} = \frac{C_{in,NAND}}{C_{inv}}
$$

In [99]:
latex_g_nand = fr"g_{{NAND_n}} = \frac{{C_{{in,NAND}}}}{{C_{{inv}}}} = \frac{{n + {Wp}}}{{{C_inv}}}"

display(Markdown("$" + latex_g_nand + "$"))
print(latex_g_nand)

$g_{NAND_n} = \frac{C_{in,NAND}}{C_{inv}} = \frac{n + 3}{4}$

g_{NAND_n} = \frac{C_{in,NAND}}{C_{inv}} = \frac{n + 3}{4}



The parasitic delay is calculated by finding the sum of output diffusion capacitances of the nMOS and pMOS transistors. 
- The nMOS transistors are in series, so only one of them contributes to the output diffusion capacitance. 
- The pMOS transistors are in parallel, so all of them contribute to the output diffusion capacitance.

$$
C_{par,NAND_n} = W_{n,NAND} + n \cdot W_{p,NAND} = n \cdot W_n + n \cdot W_p = n \cdot (W_n + W_p)
$$

$$
p_{NAND_n} = \frac{C_{NAND_n}}{C_{inv}} = \frac{n \cdot (W_n + W_p)}{W_n + W_p} = n
$$

## n-input NOR gate

A NOR gate has:

- nMOS in parallel
- pMOS in series

$$
W_{n,NOR} = W_n
$$

$$
W_{p,NOR} = n \cdot W_p
$$

Each input drives one pMOS and one nMOS gate.

$$
C_{in,NOR} = W_{n,NOR} + W_{p,NOR} = W_n + n \cdot W_p
$$

In [100]:
latex_cin_nor = fr"C_{{in,NOR}} = W_n + n \cdot W_p = {Wn} + n \cdot {Wp} = {Wn} + n \cdot {Wp}"
display(Markdown("$" + latex_cin_nor + "$"))
print(latex_cin_nor)

$C_{in,NOR} = W_n + n \cdot W_p = 1 + n \cdot 3 = 1 + n \cdot 3$

C_{in,NOR} = W_n + n \cdot W_p = 1 + n \cdot 3 = 1 + n \cdot 3


$NOR_n$ ratio 3

<img src="img/nornr3.svg" width="20%">

Example of a NOR3 with ratio 3.

$W_{n,NOR_3} = W_n = 1$

$W_{p,NOR_3} = 3 \cdot W_p = 3 \cdot 3 = 9$.

<img src="img/nor3r3.svg" width="20%">

The logical effort is

$$
g_{NOR_{n}} = \frac{C_{in,NOR}}{C_{inv}}
$$

In [101]:
latex_g_nor = fr"g_{{NOR_n}} = \frac{{C_{{in,NOR}}}}{{C_{{inv}}}} = \frac{{{Wn} + n \cdot {Wp}}}{{{C_inv}}}"

display(Markdown("$" + latex_g_nor + "$"))
print(latex_g_nor)

$g_{NOR_n} = \frac{C_{in,NOR}}{C_{inv}} = \frac{1 + n \cdot 3}{4}$

g_{NOR_n} = \frac{C_{in,NOR}}{C_{inv}} = \frac{1 + n \cdot 3}{4}


The parasitic delay is calculated by finding the sum of output diffusion capacitances of the nMOS and pMOS transistors. 

- The nMOS transistors are in parallel, so all of them contribute to the output diffusion capacitance. 
- The pMOS transistors are in series, so only one of them contributes to the output diffusion capacitance.

$$
C_{par,NOR_n} = n \cdot W_{n,NOR} + W_{p,NOR}
$$

$$
p_{NOR_n} = \frac{C_{NOR_n}}{C_{inv}} = \frac{n \cdot W_n + n \cdot W_p}{W_n + W_p} = n
$$

In [112]:
# Standard input-load table for INV, NANDn, NORn (n = 2..8)

n_values = list(range(2, 9))

inv_values = [C_inv] + ["-"] * 5
nand_values = [k * Wn + Wp for k in n_values]
nor_values = [Wn + k * Wp for k in n_values]

header = "| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |"
sep = "|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|"

rows = [
    f"| $C_{{inv}}$ | {Wn + Wp} {' | -' * 7} | ${Wn} + {Wp}$ |",
    f"| $C_{{in,NAND_n}}$ | - | {' | '.join(map(str, nand_values))} | $n\\cdot {Wn}+{Wp}$ |",
    f"| $C_{{in,NOR_n}}$ | - | {' | '.join(map(str, nor_values))} | ${Wn}+n\\cdot {Wp}$ |",
]

md_table_std = "\n".join([header, sep] + rows)
display(Markdown(md_table_std))
print(md_table_std)

| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| $C_{inv}$ | 4  | - | - | - | - | - | - | - | $1 + 3$ |
| $C_{in,NAND_n}$ | - | 5 | 6 | 7 | 8 | 9 | 10 | 11 | $n\cdot 1+3$ |
| $C_{in,NOR_n}$ | - | 7 | 10 | 13 | 16 | 19 | 22 | 25 | $1+n\cdot 3$ |

| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| $C_{inv}$ | 4  | - | - | - | - | - | - | - | $1 + 3$ |
| $C_{in,NAND_n}$ | - | 5 | 6 | 7 | 8 | 9 | 10 | 11 | $n\cdot 1+3$ |
| $C_{in,NOR_n}$ | - | 7 | 10 | 13 | 16 | 19 | 22 | 25 | $1+n\cdot 3$ |


In [113]:
# Logical effort table for INV, NANDn, NORn (n = 2..8)

g_nand_values = [round((k * Wn + Wp) / C_inv, 2) for k in n_values]
g_nor_values = [round((Wn + k * Wp) / C_inv, 2) for k in n_values]

rows_g = [
    f"| $g_{{inv}}$ | 1 {' | -' * 7} | $1$ |",
    f"| $g_{{NAND_n}}$ | - | {' | '.join(map(str, g_nand_values))} | $\\frac{{n\\cdot {Wn}+{Wp}}}{{{C_inv}}}$ |",
    f"| $g_{{NOR_n}}$ | - | {' | '.join(map(str, g_nor_values))} | $\\frac{{{Wn}+n\\cdot {Wp}}}{{{C_inv}}}$ |",
]

md_table_g = "\n".join([header, sep] + rows_g)
display(Markdown(md_table_g))
print(md_table_g)

| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| $g_{inv}$ | 1  | - | - | - | - | - | - | - | $1$ |
| $g_{NAND_n}$ | - | 1.25 | 1.5 | 1.75 | 2.0 | 2.25 | 2.5 | 2.75 | $\frac{n\cdot 1+3}{4}$ |
| $g_{NOR_n}$ | - | 1.75 | 2.5 | 3.25 | 4.0 | 4.75 | 5.5 | 6.25 | $\frac{1+n\cdot 3}{4}$ |

| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| $g_{inv}$ | 1  | - | - | - | - | - | - | - | $1$ |
| $g_{NAND_n}$ | - | 1.25 | 1.5 | 1.75 | 2.0 | 2.25 | 2.5 | 2.75 | $\frac{n\cdot 1+3}{4}$ |
| $g_{NOR_n}$ | - | 1.75 | 2.5 | 3.25 | 4.0 | 4.75 | 5.5 | 6.25 | $\frac{1+n\cdot 3}{4}$ |


# Rise and fall times

For the specific exercise we are given:

$R_{eqn,\square} = 750 \, \Omega/\square$ and $R_{eqp,\square} = 5000 \,\Omega/\square$.

To calculate the rise and fall times we can use the formula:

$$\frac{t_r}{t_f} = \frac{R_{eqp,\square}}{B_0 \cdot R_{eqn,\square}} = \frac{5000}{3 \cdot 750} = 2.22$$



In [103]:
R_eqn_P = 750
R_eqp_P = 5000

t_ratio = R_eqp_P / (B0 * R_eqn_P)

latex_t_ratio = fr"\frac{{t_r}}{{t_f}} = \frac{{R_{{eqp,\square}}}}{{B_0 \cdot R_{{eqn,\square}}}} = \frac{{{R_eqp_P}}}{{{B0} \cdot {R_eqn_P}}} = {t_ratio}"
display(Markdown("$" + latex_t_ratio + "$"))
print(latex_t_ratio)

$\frac{t_r}{t_f} = \frac{R_{eqp,\square}}{B_0 \cdot R_{eqn,\square}} = \frac{5000}{3 \cdot 750} = 2.2222222222222223$

\frac{t_r}{t_f} = \frac{R_{eqp,\square}}{B_0 \cdot R_{eqn,\square}} = \frac{5000}{3 \cdot 750} = 2.2222222222222223


## Minimum delay in units of $\tau$

Lecture 4 slide 52

We are given best stage effort

$\hat{f} = 7.5$

The best stage effort is defined as $\hat{f} = F^{\frac{1}{N}}$ where $F$ is path effort and $N$ is the number of stages.

The minimum delay in units of $\tau$ is given by the formula:

$$D = N \cdot F^{\frac{1}{N}} + P$$

From earlier we calculated the parasitic delay for the gates to be 

$$p_{NAND_n} = n$$
$$p_{NOR_n} = n$$


For the assigtnment, the critical path goes through NAND2, NAND2, NOR3, NOR3.

$$P = 2 + 2 + 3 + 3 = 10$$

The least delay is:

$$
D = N \cdot F^{\frac{1}{N}} + P = 4 \cdot 7.5 + 10 = 40 \, \tau
$$

# Determining gate sizes

Lecture 4 slide 44
Book around page 165

When sizing for best effort we can use the formula:

$$
C_{in,i}=\frac{C_{out,i}\,g_i\,b_i}{\hat{f}}
$$

We calculate branching factors for each stage:

$$b_i=\frac{C_{\text{total driven by stage }i}}{C_{\text{on-path load}}}$$

can also be written as:

$$b_i=\frac{C_{\text{on-path}}+C_{\text{off-path}}}{C_{\text{on-path}}}$$

For this specific task we have the path NAND2, NAND2, NOR3, NOR3 with ratio 3 and output load of 9.

We calculated g for ratio 3 to be:

| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| $g_{inv}$ | 1  | - | - | - | - | - | - | - | $1$ |
| $g_{NAND_n}$ | - | 1.25 | 1.5 | 1.75 | 2.0 | 2.25 | 2.5 | 2.75 | $\frac{n\cdot 1+3}{4}$ |
| $g_{NOR_n}$ | - | 1.75 | 2.5 | 3.25 | 4.0 | 4.75 | 5.5 | 6.25 | $\frac{1+n\cdot 3}{4}$ |

We calculated the input loads for ratio 3 to be:

| Gate | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | n |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| $C_{inv}$ | 4  | - | - | - | - | - | - | - | $1 + 3$ |
| $C_{in,NAND_n}$ | - | 5 | 6 | 7 | 8 | 9 | 10 | 11 | $n\cdot 1+3$ |
| $C_{in,NOR_n}$ | - | 7 | 10 | 13 | 16 | 19 | 22 | 25 | $1+n\cdot 3$ |


### Stage 4 (last stage)

Gate is NOR3, so $g_{NOR_3} = 2.5$

$C_{out,4} = 9$

No branching for the last stage
$$b_4 = 1$$

$$
C_{in,4} = \frac{C_{out,4} \cdot g_{NOR_3} \cdot b_4}{\hat{f}} = \frac{9 \cdot 2.5 \cdot 1}{7.5} = 3
$$

The input load for a NOR3 with ratio 3 is 10, so we need to scale the gate by a factor of $\frac{3}{10} = 0.3$ to get the correct input load.

### Stage 3

Gate is NOR3, so $g_{NOR_3} = 2.5$

$C_{out,3} = C_{in,4} = 3$

We are branching out. 

$$b_3 = \frac{C_{in,NOR_3} + C_{in,NOR_4} + C_{in,NAND_4}}{C_{in,NOR_3}} = \frac{10 + 13 + 7}{10} = \frac{30}{10} = 3$$

$$C_{in,3} = \frac{C_{in,4} \cdot g_{NOR_3} \cdot b_3}{7.5} = \frac{3 \cdot 2.5 \cdot 3}{7.5} = 3$$

The input load for a NOR3 with ratio 3 is 10, so we need to scale the gate by a factor of $\frac{3}{10} = 0.3$ to get the correct input load.

### Stage 2

Gate is NAND2, so $g_{NAND_2} = 1.25$

$C_{out,2} = C_{in,3} = 3$

We are branching out. Branching looks the same as we have the same gates and same loads as in stage 3.

$$b_2 = \frac{C_{in,NOR_3} + C_{in,NOR_4} + C_{in,NAND_4}}{C_{in,NOR_3}} = \frac{10 + 13 + 7}{10} = \frac{30}{10} = 3$$

$$C_{in,2} = \frac{C_{in,3} \cdot g_{NAND_2} \cdot b_2}{\hat{f}} = \frac{3 \cdot 1.25 \cdot 3}{7.5} = 1.5$$

The input load for a NAND2 with ratio 3 is 5, so we need to scale the gate by a factor of $\frac{1.5}{5} = 0.3$ to get the correct input load.

### Stage 1

Gate is NAND2, so $g_{NAND_2} = 1.25$

$C_{out,1} = C_{in,2} = 1.5$

We are branching out. 

$$C_{in,NOR_6} = 1 + 6 \cdot 3 = 19$$
$$C_{in,NAND_2} = 2 + 3 = 5$$
$$C_{in,NOR_5} = 1 + 5 \cdot 3 = 16$$

$$b_1 = \frac{C_{in,NOR_6} + C_{in,NAND_2} + C_{in,NOR_5}}{C_{in,NAND_2}} = \frac{19 + 5 + 16}{5} = \frac{40}{5} = 8$$

$$C_{in,1} = \frac{C_{out,1} \cdot g_{NAND_2} \cdot b_1}{\hat{f}} = \frac{1.5 \cdot 1.25 \cdot 8}{7.5} = 2$$

The input load for a NAND2 with ratio 3 is 5, so we need to scale the gate by a factor of $\frac{2}{5} = 0.4$ to get the correct input load.

#  Path electrical effort, H

Assuming that it is allowed to scale down the gates, we can calculate the path electrical effort as:

$$
H = \frac{C_{out}}{C_{in,1}} = \frac{9}{2} = 4.5
$$

# Questions

- Should I calculate branching factor based on the standard input load of each gate?
- If I calculate a new $C_{in,4} = 3$ how will this affect my branching calculation? If so, should I still consider the off-path branches to have standard input load?
- Is it expected to have a scaling factor below 1 for the gates? How should I deal with scaling down a gate when nmos is already using the minimum width?